sys/net/network_layer/ng_ndp/ng_ndp.c

```diff
@@ -0,0 +1,950 @@
+            break;
+    }
+}
+
+void ng_ndp_netif_add(ng_ipv6_netif_t *iface)
+{
+    uint32_t reach_time = _rand(NG_NDP_MIN_RAND, NG_NDP_MAX_RAND);
+
+    /* set default values */
+    mutex_lock(&iface->mutex);
+    iface->reach_time_base = NG_NDP_REACH_TIME;
+    reach_time = (reach_time * iface->reach_time_base) / 10;
+    iface->reach_time = timex_set(0, reach_time);
+    timex_normalize(&iface->reach_time);
+    iface->retrans_timer = timex_set(0, NG_NDP_RETRANS_TIMER);
+    timex_normalize(&iface->retrans_timer);
+    mutex_unlock(&iface->mutex);
+}
+
+void ng_ndp_netif_remove(ng_ipv6_netif_t *iface)
+{
+    /* TODO */
+}
+
+static ng_ipv6_addr_t *_default_router(void)
+{
+    ng_ipv6_nc_t *router = ng_ipv6_nc_get_next_router(NULL);
+
+    /* first look if there is any reachable router */
+    while (router != NULL) {
+        if ((ng_ipv6_nc_get_state(router) != NG_IPV6_NC_STATE_INCOMPLETE) &&
+            (ng_ipv6_nc_get_state(router) != NG_IPV6_NC_STATE_UNREACHABLE)) {
+            _last_router = NULL;
+
+            return &router->ipv6_addr;
+        }
+
+        router = ng_ipv6_nc_get_next_router(router);
+    }
+
+    /* else take the first one, but keep round-robin in further selections */
+    router = ng_ipv6_nc_get_next_router(_last_router);
+
+    if (router == NULL) {   /* end of router list or there is none => wrap around */
+        router = ng_ipv6_nc_get_next_router(router);
+
+        if (router == NULL) {   /* still nothing found => no router in list */
+            return NULL;
+        }
+    }
+
+    _last_router = router;
+
+    return &router->ipv6_addr;
+}
+
+kernel_pid_t ng_ndp_next_hop_l2addr(uint8_t *l2addr, uint8_t *l2addr_len,
+                                    kernel_pid_t iface, ng_ipv6_addr_t *dst,
+                                    ng_pktsnip_t *pkt)
+{
+    ng_ipv6_addr_t *next_hop_ip = NULL, *prefix = NULL;
+#ifdef MODULE_NG_IPV6_ROUTER
+    size_t next_hop_size;
+
+    if ((fib_get_next_hop(&iface, (uint8_t *)next_hop_ip, &next_hop_size,
+                          (uint8_t *)dst, sizeof(ng_ipv6_addr_t),
+                          0) < 0) || (next_hop_ip != sizeof(ng_ipv6_addr_t))) {
+        next_hop_ip = NULL;
+    }
+#endif
+
+    if ((next_hop_ip == NULL)) {            /* no route to host */
+        if (iface == KERNEL_PID_UNDEF) {
+            /* ng_ipv6_netif_t doubles as prefix list */
+            iface = ng_ipv6_netif_find_by_prefix(&prefix, dst);
+        }
+        else {
+            /* ng_ipv6_netif_t doubles as prefix list */
+            prefix = ng_ipv6_netif_match_prefix(iface, dst);
+        }
+
+        if ((prefix != NULL) &&             /* prefix is on-link */
+            (ng_ipv6_netif_addr_get(prefix)->flags &
+             NG_IPV6_NETIF_ADDR_FLAGS_NDP_ON_LINK
+            )
+           ) {
+            next_hop_ip = (ng_ipv6_addr_t *)dst;
+        }
+    }
+
+#ifndef MODULE_NG_IPV6_ROUTER
+    if (next_hop_ip == NULL) {
+        next_hop_ip = _default_router();
```

Comment:
[RFC4861](https://tools.ietf.org/html/rfc4861#section-6.3.4) paints a different picture: any interface that sends Router Solicitations should not be flagged as router (and thus removing it from the default router list). [This is not the case for 6LoWPAN optimized networks however, see [RFC 6775](https://tools.ietf.org/html/rfc6775)]. Usually a router has a FIB, resulting in the default route already being selected in l341 and this case being skipped anyway. Sure I can leave this case in, but I thought maybe we optimize here a little ;-)